src/Common/MemoryTracker.cpp

```diff
@@ -82,6 +83,29 @@ inline std::string_view toDescription(OvercommitResult result)
     }
 }
 
+bool shouldTrackAllocation(DB::Float64 probability, void * ptr)
+{
+    return intHash64(uintptr_t(ptr)) < std::numeric_limits<uint64_t>::max() * probability;
```

Comment:
I still see some minor issues here:
* the set of addresses that could be tracked is fixed. so some addresses will never be traced. although it shouldn't be a problem, because allocation in different parts of the code could be considered a random process
* how good `intHash64` distributes addresses in [0, 2^63) is also questionable
* we pay hash cost on each allocation (it is minor, because sampling is turned off by default)

what I consider as not minor is that ideally sampled values should be similar to the source distribution (i.e. as a minimum if we allocated 10GB and sampling prob is 0.1 we should collect traces that sum up in ~1GB). so I think the right approach is when prob depends on the size of allocation. as a bonus it provides that we shift more of tracing overhead towards big allocations. [here](https://github.com/jemalloc/jemalloc/blob/dev/doc_internal/PROFILING_INTERNALS.md) is how it is implemented in jemalloc. wdyt about implementing their "sampling per-byte"?

Reply:
Yes, it's a very nice idea!
Let's implement sampling per-byte (at least under flag).
But I would rather do it in a separate PR. This one is too hard to merge :)